### PR TITLE
Preserve shortcodes in URLs processed for GA tracking

### PR DESCRIPTION
**Pull request by costasovo**
### Problem
When GA tracking parameters were added to internal links containing shortcodes (e.g., [subscriber:email]), WordPress's add_query_arg() function would URL-encode the shortcodes, breaking them.

### Solution
Implemented a three-step process to preserve shortcodes:

1. Extract - Find all shortcodes in URL query parameters and replace them with safe placeholders
2. Add GA params - Use add_query_arg() to add tracking parameters (placeholders remain unencoded)
3. Restore - Replace placeholders with original shortcodes

The solution properly handles:

- Simple shortcodes: [subscriber:email]
- Complex shortcodes with defaults: [subscriber:firstname|default:Guest User]
- Shortcodes in array parameters: filters[email]=[subscriber:email]

## Code review notes
N/A

## QA notes

1. Add a link to an email paragraph. The link have to link to the site you're sending from http://yourtestsite.com/?email=[subscriber:email]&name=[subscriber:firstname|default:reader]
2. Send the email, with full tracking enabled. Use regular sending not preview
3. Verify that after clicking on the link in the url goes to something like http://yourtestsite.com/?email=admin@example.com&name=reader 

## Linked PRs
N/A

## Linked tickets
Closes STOMAIL-7665

## After-merge notes
N/A

## Summary by CodeRabbit

* **Bug Fixes**
  * Preserve shortcodes in URLs that link to the sending site when Google Analytics parameters are added, preventing them from being broken or encoded.

* **Tests**
  * Added tests covering preservation across internal links, complex shortcodes, array query parameters, and edge-case scenarios.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:stomail-7665-shortcodes-in-urls-dont-work)

_The latest successful build from `stomail-7665-shortcodes-in-urls-dont-work` will be used. If none is available, the link won't work._

### CODE REVIEW

File: mailpoet/lib/Statistics/GATracking.php [hunk 1]
```diff
@@ -108,4 +117,81 @@ private function addParams($extractedLinks, $gaCampaign, $internalHost = null) {
...
+        if (preg_match_all($pattern, $value, $matches)) {
+          foreach ($matches[0] as $shortcode) {
+            // Create a unique placeholder
+            $placeholder = 'MPSHORTCODE' . $index . 'MPEND';
+            $shortcodeMap[$placeholder] = $shortcode;
+            // Replace shortcode with placeholder directly in the URL
+            $urlWithPlaceholders = str_replace($shortcode, $placeholder, $urlWithPlaceholders);
+            $index++;
+          }
+        }
```
[the remark this answers is not in the document]
Comment (costasovo) [reply mailpoet/lib/Statistics/GATracking.php]: We don't need to handle False positives, because this logic is handled in a different layer. Here we need to preserve the syntax. The ReDoS and tests are addressed in https://github.com/mailpoet/mailpoet/pull/6443/commits/686742a950decb7e6adeb5ebc19073435280bda1 and https://github.com/mailpoet/mailpoet/pull/6443/commits/efff25136f6295a0b70f5d8310b0847980104d83

File: mailpoet/changelog/2025-12-16-12-18-50-fixed-shortcodes-in-urls-linking-the-sending-site-are-br.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# Type: Fixed
+
+# Description
+
+Shortcodes in URLs linking the sending site are broken
```
Comment (pavel-mailpoet) [inline mailpoet/changelog/2025-12-16-12-18-50-fixed-shortcodes-in-urls-linking-the-sending-site-are-br.md]: I am not sure if this is understandable to someone who doesn't know anything about this issue. Could we use something more descriptive, like "Fixed shortcodes in URL parameters getting broken when GA tracking is enabled"

**Review by pavel-mailpoet — APPROVED** [review]
Looks good to me. I have just one small suggestion: either update it or not, then pass it to QA.

**Comment by costasovo** [comment]
@pavel-mailpoet Thank you for the fast review. I've updated the changelog.

**Review by Aschepikov — APPROVED** [review]
GJ!
